### PR TITLE
make some windows process and shellcraft

### LINKED ISSUE #1930: launch processes on windows
* What the feature does

Integrates the pwintools debugger to pwntools See: https://github.com/masthoon/pwintools

* Why the feature should exist

As told in this issue : https://github.com/masthoon/pwintools/issues/4 pwintools should have been written in upstream of pwntools. I want to remediate to this issue.

* What tests should be included

Some tests about processs.

I will make a pull request. Assign me to this issue please.

### PR DESCRIPTION
**Pull request by gogo2464**
Closes #1930. duplicate of https://github.com/Gallopsled/pwntools/pull/1931. I do not know which is better.
A new vm for travis should be added. I will write some tests in the future.

# Pwntools Pull Request

Thanks for contributing to Pwntools!  Take a moment to look at [`CONTRIBUTING.md`][contributing] to make sure you're familiar with Pwntools development.

Please provide a high-level explanation of what this pull request is for.

## Testing

Pull Requests that introduce new code should try to add doctests for that code.  See [`TESTING.md`][testing] for more information.

## Target Branch

Depending on what the PR is for, it needs to target a different branch.

You can always [change the branch][change] after you create the PR if it's against the wrong branch.

| Branch   | Type of PR                                                       |
| -------- | ---------------------------------------------------------------- |
| `dev`    | New features, and enhancements
| `dev`    | Documentation fixes and new tests
| `stable` | Bug fixes that affect the current `stable` branch
| `beta`   | Bug fixes that affect the current `beta` branch, but not `stable`
| `dev`    | Bug fixes for code that has never been released

[contributing]: https://github.com/Gallopsled/pwntools/blob/dev/CONTRIBUTING.md
[testing]: https://github.com/Gallopsled/pwntools/blob/dev/TESTING.md
[change]: https://github.com/blog/2224-change-the-base-branch-of-a-pull-request

## Changelog

After creating your Pull Request, please add and push a commit that updates the changelog for the appropriate branch.  
You can look at the existing changelog for examples of how to do this.

### CODE REVIEW

**Comment by heapcrash** [comment]
To merge this, there needs to be some way of testing the Windows changes.

Please add a custom Github Actions for Windows step / action to run whatever test harness you create to show this works.

You will also need to update the existing Github Actions to ensure they do NOT run on Windows, since the doctests will certainly fail.

File: pwnlib/tubes/__init__.py [hunk 1]
```diff
@@ -12,12 +12,12 @@
 from __future__ import absolute_import
 
 from pwnlib.tubes import listen
-from pwnlib.tubes import process
+from pwnlib.tubes import os_process
```
Comment (heapcrash) [inline pwnlib/tubes/__init__.py:15]: Do not change the module name.

File: pwnlib/tubes/os_process/process_linux.py [hunk 2]
```diff
@@ -0,0 +1,977 @@
+# -*- coding: utf-8 -*-
```
Comment (heapcrash) [inline pwnlib/tubes/os_process/process_linux.py:1]: It looks like Git did not detect this as a rename.  I recommend re-starting the PR with a rename, e.g.

```
git mv process.py process_linux.py
```

File: pwnlib/tubes/process.py [hunk 3]
```diff
@@ -1,1034 +1,6 @@
-# -*- coding: utf-8 -*-
```
Comment (heapcrash) [inline pwnlib/tubes/process.py:1]: Use `git mv` do not simply copy the data.

File: pwnlib/tubes/os_process/process_windows.py [hunk 4]
```diff
@@ -0,0 +1,960 @@
...
+    def disasm(data, bitness=64, vma=0):
+        """disasm(data, bitness = 64, vma = 0) dissas the data at vma"""
+        cs = capstone.Cs(capstone.CS_ARCH_X86, capstone.CS_MODE_64 if bitness == 64 else capstone.CS_MODE_32)
+        dis = ''
+        for i in cs.disasm(data, vma):
+            dis += "%x:\t%s\t%s\n" % (i.address, i.mnemonic, i.op_str)
+        return dis
+except ImportError:
+    def disasm(data, bitness=64, vma=0):
+        raise (NotImplementedError("Capstone module not found"))
```
Comment (heapcrash) [inline pwnlib/tubes/os_process/process_windows.py:36]: Why are we doing this?  We do not use `capstone` as a disassembler, and importing it on `from pwn import *` greatly increases the startup time.

Do not overload / overwrite the names of other existing functions.  This should go in `asm.py`.

File: pwnlib/tubes/os_process/process_windows.py [hunk 5]
```diff
@@ -0,0 +1,960 @@
...
+
+
+    def asm(code, bitness=64, vma=0):
+        """asm(code, bitness = 64, vma = 0) assembles the assembly code at vma"""
+        ks = keystone.Ks(keystone.KS_ARCH_X86, keystone.KS_MODE_64 if bitness == 64 else keystone.KS_MODE_32)
+        encoding, count = ks.asm(code, vma)
+        return encoding
+except ImportError:
+    def asm(code, bitness=64, vma=0):
+        raise (NotImplementedError("Keystone module not found"))
```
Comment (heapcrash) [inline pwnlib/tubes/os_process/process_windows.py:49]: This belongs in asm.py

File: setup.py [hunk 6]
```diff
@@ -65,6 +65,7 @@
                         'six>=1.12.0',
                         'rpyc',
                         'colored_traceback',
+                        'git+git://github.com/hakril/PythonForWindows',
```
Comment (heapcrash) [inline setup.py:68]: No.

**Review by heapcrash — CHANGES_REQUESTED** [review]
This is a mess.  Try with a smaller PR first.

Specifically, all of the random stuff imported in process_win is not acceptable.

I recommend closing this PR, and starting small.  For example, do not rename or modify `process.py` at all.

Add only what is absolutely minimally needed to `win_process`, it should inherit from `pwnlib.tubes.process.process()` and change whatever behavior is needed -- but it should not re-implement everything.

Do **NOT** add dependencies on completely separate repositories, ESPECIALLY by directly linking them to a Git repo.  If you need additional dependencies, they should be packages on PyPi.

Separately, there's a ton of random cruft in the process_windows.py that does not belong there.  Changes to asm, disasm, DotDict, xor_pair, and all of the other random functions you imported / redefined are completely unacceptable.  **Do not do this.**

I'm closing out this PR.  Please try again with a smaller PR.  I suggest that before you add a single line of new code, you should make it so that Pwntools will import cleanly with `from pwn import *`.

Then make your first pull request.

After that, we can talk more about NEW code for Windows-specific stuff.

**Comment by gogo2464** [comment]
I can only agree with you. I should not have done all this. Can I make another PR?

**Comment by Arusekk** [comment]
You can do it in the same PR, but use `git push -f` to the branch whenever you are ready, so that each commit does one thing and the whole can be reviewed easier.

**Comment by gogo2464** [comment]
Cool. Thank you!

**Comment by gogo2464** [comment]
-I still have the requirement issue because I am looking for the good version for pip.

-I removed the useless code. Some usefull can be added about shellcraft but later and `incrementally`.

-I did not have added the travis files yet.


You can still get a look.

**Pull request closed without merging**